File: samlang-checker/src/test/kotlin/samlang/checker/GlobalTypingContextBuilderTest.kt
package samlang.checker

import io.kotlintest.shouldBe
import io.kotlintest.specs.StringSpec
import kotlinx.collections.immutable.persistentMapOf
import samlang.ast.common.ModuleMembersImport
import samlang.ast.common.ModuleReference
import samlang.ast.common.Range
import samlang.ast.common.Sources
import samlang.ast.common.TypeDefinition
import samlang.ast.lang.ClassDefinition
import samlang.ast.lang.Module
import samlang.checker.GlobalTypingContext.ClassType
import samlang.checker.GlobalTypingContext.ModuleTypingContext

class GlobalTypingContextBuilderTest : StringSpec() {
    init {
        "can handle imports and definitions" {
            val module0Reference = ModuleReference(moduleName = "Module0")
            val module1Reference = ModuleReference(moduleName = "Module1")
            val typeDefinition = TypeDefinition.ofDummy()
            val class0 = ClassDefinition(
                range = Range.DUMMY,
                name = "Class0",
                nameRange = Range.DUMMY,
                typeDefinition = typeDefinition,
                members = emptyList()
            )
            val class1 = ClassDefinition(
                range = Range.DUMMY,
                name = "Class1",
                nameRange = Range.DUMMY,
                typeDefinition = typeDefinition,
                members = emptyList()
            )
            val module0 = Module(imports = emptyList(), classDefinitions = listOf(class0))
            val module1 = Module(
                imports = listOf(
                    ModuleMembersImport(
                        range = Range.DUMMY,
                        importedModuleRange = Range.DUMMY,
                        importedModule = module0Reference,
                        importedMembers = listOf("Class0" to Range.DUMMY)
                    )
                ),
                classDefinitions = listOf(class1)
            )
            val sources = Sources(moduleMappings = mapOf(module0Reference to module0, module1Reference to module1))
            val commonClassType = ClassType(
                typeDefinition = typeDefinition, methods = persistentMapOf(), functions = persistentMapOf()
            )
            GlobalTypingContextBuilder.buildGlobalTypingContext(sources = sources) shouldBe GlobalTypingContext(
                modules = persistentMapOf(
                    module0Reference to ModuleTypingContext(classes = persistentMapOf("Class0" to commonClassType)),
                    module1Reference to ModuleTypingContext(
                        classes = persistentMapOf("Class0" to commonClassType, "Class1" to commonClassType)
                    )
                )
            )
        }
    }
}

File: samlang-core-checker/__tests__/global-typing-context-builder.test.ts
import { functionType, intType, Range, ModuleReference } from 'samlang-core-ast/common-nodes';
import { SourceExpressionFalse } from 'samlang-core-ast/samlang-nodes';
import type { SourceClassDefinition, SamlangModule } from 'samlang-core-ast/samlang-nodes';
import { mapOf } from 'samlang-core/utils';

import {
  buildGlobalTypingContext,
  updateGlobalTypingContext,
} from '../global-typing-context-builder';

const module0Reference = new ModuleReference(['Module0']);
const module1Reference = new ModuleReference(['Module1']);

const typeDefinition = {
  range: Range.DUMMY,
  type: 'object',
  names: [],
  mappings: {},
} as const;

const class0: SourceClassDefinition = {
  range: Range.DUMMY,
  associatedComments: [],
  name: 'Class0',
  nameRange: Range.DUMMY,
  typeParameters: [],
  typeDefinition,
  members: [],
};
const class1: SourceClassDefinition = {
  range: Range.DUMMY,
  associatedComments: [],
  name: 'Class1',
  nameRange: Range.DUMMY,
  typeParameters: [],
  typeDefinition,
  members: [
    {
      associatedComments: [],
      range: Range.DUMMY,
      isPublic: true,
      isMethod: true,
      nameRange: Range.DUMMY,
      name: 'm1',
      typeParameters: [],
      type: functionType([], intType),
      parameters: [],
      body: SourceExpressionFalse(),
    },
    {
      associatedComments: [],
      range: Range.DUMMY,
      isPublic: false,
      isMethod: false,
      nameRange: Range.DUMMY,
      name: 'f1',
      typeParameters: [],
      type: functionType([], intType),
      parameters: [],
      body: SourceExpressionFalse(),
    },
  ],
};
const class2: SourceClassDefinition = {
  range: Range.DUMMY,
  associatedComments: [],
  name: 'Class2',
  nameRange: Range.DUMMY,
  typeParameters: [],
  typeDefinition,
  members: [],
};

const module0: SamlangModule = { imports: [], classes: [class0] };
const module1: SamlangModule = {
  imports: [
    {
      range: Range.DUMMY,
      importedModule: module0Reference,
      importedModuleRange: Range.DUMMY,
      importedMembers: [
        ['Class0', Range.DUMMY],
        ['BAD_CLASS_THAT_DOESNT_EXIST', Range.DUMMY],
      ],
    },
  ],
  classes: [class1, class2],
};

const testSources = mapOf([module0Reference, module0], [module1Reference, module1]);

describe('global-typing-context-builder', () => {
  it('can handle imports and definitions', () => {
    const actualGlobalTypingContext = buildGlobalTypingContext(testSources, {});
    expect(actualGlobalTypingContext.size).toBe(3);

    expect(actualGlobalTypingContext.get(module0Reference)).toStrictEqual({
      Class0: { typeParameters: [], typeDefinition, functions: {}, methods: {} },
    });
    expect(actualGlobalTypingContext.get(module1Reference)).toStrictEqual({
      Class1: {
        typeParameters: [],
        typeDefinition,
        functions: {
          f1: { isPublic: false, type: functionType([], intType), typeParameters: [] },
        },
        methods: {
          m1: { isPublic: true, type: functionType([], intType), typeParameters: [] },
        },
      },
      Class2: {
        typeParameters: [],
        typeDefinition,
        functions: {},
        methods: {},
      },
    });
  });

  it('can handle incremental add', () => {
    const actualGlobalTypingContext = buildGlobalTypingContext(testSources, {});
    updateGlobalTypingContext(
      actualGlobalTypingContext,
      mapOf(
        [module0Reference, module0],
        [module1Reference, { ...module1, classes: [class1, class2] }]
      ),
      [module0Reference, module1Reference]
    );

    expect(actualGlobalTypingContext.size).toBe(3);

    expect(actualGlobalTypingContext.get(module0Reference)).toStrictEqual({
      Class0: { typeParameters: [], typeDefinition, functions: {}, methods: {} },
    });
    expect(actualGlobalTypingContext.get(module1Reference)).toStrictEqual({
      Class1: {
        typeParameters: [],
        typeDefinition,
        functions: {
          f1: { isPublic: false, type: functionType([], intType), typeParameters: [] },
        },
        methods: {
          m1: { isPublic: true, type: functionType([], intType), typeParameters: [] },
        },
      },
      Class2: { typeParameters: [], typeDefinition, functions: {}, methods: {} },
    });
  });

  it('can handle incremental update', () => {
    const actualGlobalTypingContext = buildGlobalTypingContext(testSources, {});
    updateGlobalTypingContext(
      actualGlobalTypingContext,
      mapOf([module0Reference, { imports: [], classes: [] }], [module1Reference, module1]),
      [module0Reference, module1Reference]
    );

    expect(actualGlobalTypingContext.size).toBe(3);

    expect(actualGlobalTypingContext.get(module0Reference)).toStrictEqual({});
    expect(actualGlobalTypingContext.get(module1Reference)).toStrictEqual({
      Class1: {
        typeParameters: [],
        typeDefinition,
        functions: {
          f1: { isPublic: false, type: functionType([], intType), typeParameters: [] },
        },
        methods: {
          m1: { isPublic: true, type: functionType([], intType), typeParameters: [] },
        },
      },
      Class2: {
        typeParameters: [],
        typeDefinition,
        functions: {},
        methods: {},
      },
    });
  });

  it('can handle incremental removal', () => {
    const actualGlobalTypingContext = buildGlobalTypingContext(testSources, {});
    updateGlobalTypingContext(actualGlobalTypingContext, mapOf([module1Reference, module1]), [
      module0Reference,
      module1Reference,
    ]);
    expect(actualGlobalTypingContext.size).toBe(2);
  });
});
